include/pybind11/pybind11.h

```diff
@@ -468,13 +474,20 @@ class cpp_function : public function {
         if (rec->sibling) {
             if (PyCFunction_Check(rec->sibling.ptr())) {
                 auto *self = PyCFunction_GET_SELF(rec->sibling.ptr());
-                capsule rec_capsule = isinstance<capsule>(self) ? reinterpret_borrow<capsule>(self)
-                                                                : capsule(self);
-                chain = (detail::function_record *) rec_capsule;
-                /* Never append a method to an overload chain of a parent class;
-                   instead, hide the parent's overloads in this case */
-                if (!chain->scope.is(rec->scope)) {
+                if (!isinstance<capsule>(self)) {
                     chain = nullptr;
+                } else {
+                    auto rec_capsule = reinterpret_borrow<capsule>(self);
+                    if (rec_capsule.name() == detail::function_capsule_name()) {
```

Comment:
Is it really safe to do pointer comparison here? Addresses from different submodules could be different? Or maybe if a capsule from a different module is somehow added? Should be std::strcmp, right?

Reply:
The pointer comparison is on purpose to avoid binary compatibility issues. Otherwise we would have multiple detail::function_record's with different ABIs trying to interact with each other (which could be a disaster if they are from different compilers / STL versions).

I can switch it to std::strcmp, but that is much less safe.

Reply:
I am a bit confused as to why it's less safe. Are we sure a foreign capsule could never be a self object here? I think we may be able to construct a case where it is? (Ie. a pybind11 class the subclasses an imported pybind11 class, right?)

Reply:
> Even if the function is not constexpr, I think that variable can be a static constexpr const char *

static constexpr const char* and my current function have different semantics.

static constexpr const char* creates a single variable per translation unit.

My current function creates a single actually global variable that is shared across an *.so.

We want to ensure that things are shared across an *.so so we can't use a static variable here.

Reply:
> I am a bit confused as to why it's less safe. Are we sure a foreign capsule could never be a self object here? I think we may be able to construct a case where it is? (Ie. a pybind11 class the subclasses an imported pybind11 class, right?)

The case where it would be less safe is right here in functional.h 

If you pass a capsule function from a foreign pybind11 class, it will break.

Reply:
> Ie. a pybind11 class the subclasses an imported pybind11 class, right?

This is undefined behavior unless both classes are on the exact same ABI (same compiler, same STL, same pybind11 version, etc).

Reply:
We should add a comment for these comparisons that these are pointer comparisons and not string comparison bugs.